--- grass/raster/r.out.gdal/export_band.c	2014-12-19 21:20:34.724904140 +0100
+++ grass/raster/r.out.gdal/export_band.c	2014-12-19 21:20:37.214922097 +0100
@@ -236,11 +236,6 @@
 	Rast_get_fp_range_min_max(&sRange, &dfCellMin, &dfCellMax);
     }
 
-    /* suppress useless warnings */
-    CPLPushErrorHandler(CPLQuietErrorHandler);
-    GDALSetRasterColorInterpretation(hBand, GPI_RGB);
-    CPLPopErrorHandler();
-    
     sprintf(value, "GRASS GIS %s", GRASS_VERSION_NUMBER);
     GDALSetMetadataItem(hBand, "Generated_with", value, NULL);
 
